File: clients/ts/formal/quint-itf-replay.mjs
#!/usr/bin/env node

/**
 * Replay Quint Informal Trace Format (ITF) model-based traces against the
 * production TypeScript OptoSyncClient and its Dexie/IndexedDB queue.
 *
 * The model owns server-side ledger/commit facts. This adapter synthesizes only
 * those responses, while mutation allocation, durable queue insertion, immutable
 * request construction, acknowledgement rejection/application, checkpointing,
 * and snapshot installation execute through the compiled client package.
 */

import 'fake-indexeddb/auto';

import { readFile } from 'node:fs/promises';
import { createRequire } from 'node:module';
import path from 'node:path';
import process from 'node:process';

const require = createRequire(import.meta.url);
const { OptoSyncClient, SYNC_STATUS } = require('../dist/client.js');

const MUTATION_SEQUENCE_KEY = 'mutation.seq';
let databaseSequence = 0;

function invalid(message) {
  const error = new Error(message);
  error.name = 'FormalReplayError';
  return error;
}

function ensure(condition, message) {
  if (!condition) throw invalid(message);
}

function field(value, name) {
  ensure(
    value !== null && typeof value === 'object' && !Array.isArray(value),
    `expected object while reading ITF field \`${name}\``,
  );
  ensure(
    Object.prototype.hasOwnProperty.call(value, name),
    `missing ITF field \`${name}\``,
  );
  return value[name];
}

function taggedBigInt(value) {
  const encoded = field(value, '#bigint');
  ensure(typeof encoded === 'string', 'ITF #bigint must contain a decimal string');
  ensure(/^(?:0|[1-9]\d*)$/.test(encoded), `invalid ITF #bigint \`${encoded}\``);
  return BigInt(encoded);
}

function stateBigInt(state, name) {
  return taggedBigInt(field(state.s, name));
}

function stateBool(state, name) {
  const value = field(state.s, name);
  ensure(typeof value === 'boolean', `ITF state field \`${name}\` must be boolean`);
  return value;
}

function stateSet(state, name) {
  const entries = field(field(state.s, name), '#set');
  ensure(Array.isArray(entries), `ITF state field \`${name}\` must be a set`);
  return new Set(entries.map((entry) => taggedBigInt(entry).toString()));
}

function stateTag(state, name) {
  const tag = field(field(state.s, name), 'tag');
  ensure(typeof tag === 'string', `ITF state field \`${name}\` must be tagged`);
  return tag;
}

function pickedId(state) {
  const pick = field(state.nondetPicks ?? {}, 'id');
  ensure(
    field(pick, 'tag') === 'Some',
    `action \`${state.action}\` requires a nondeterministic id`,
  );
  return taggedBigInt(field(pick, 'value'));
}

function requestMutationId(request) {
  ensure(request !== null, 'no in-flight request');
  ensure(
    Array.isArray(request.mutations) && request.mutations.length === 1,
    'formal adapter sends exactly one mutation per request',
  );
  const id = request.mutations[0]?.mutationId;
  ensure(typeof id === 'string' && /^[1-9]\d*$/.test(id), 'invalid request mutation id');
  return BigInt(id);
}

function sortedSet(set) {
  return [...set].sort((left, right) => {
    const a = BigInt(left);
    const b = BigInt(right);
    return a < b ? -1 : a > b ? 1 : 0;
  });
}

function ensureSetsEqual(actual, expected, context, name) {
  const left = sortedSet(actual);
  const right = sortedSet(expected);
  ensure(
    JSON.stringify(left) === JSON.stringify(right),
    `${context}: ${name} ${JSON.stringify(left)} != model ${JSON.stringify(right)}`,
  );
}

function allocatedIds(nextId) {
  const result = new Set();
  for (let id = 1n; id < nextId; id += 1n) result.add(id.toString());
  return result;
}

async function createAdapter(tracePath) {
  databaseSequence += 1;
  const safeName = path.basename(tracePath).replace(/[^A-Za-z0-9_.-]/g, '-');
  const databaseName = `OptoSyncFormal-${process.pid}-${databaseSequence}-${safeName}`;
  const client = new OptoSyncClient({
    databaseName,
    stampUpdatedAt: false,
  });
  await client.db.open();
  return {
    client,
    request: null,
    response: null,
    responseValid: false,
    replacingSnapshot: false,
  };
}

async function destroyAdapter(adapter) {
  if (!adapter) return;
  try {
    adapter.client.db.close();
    await adapter.client.db.delete();
  } catch (error) {
    process.stderr.write(`warning: failed to delete formal replay database: ${error}\n`);
  }
}

async function responseFromState(adapter, state, status, originalStatus) {
  const mutationId = stateBigInt(state, 'response_mutation_id').toString();
  const checkpoint = stateBigInt(state, 'response_checkpoint').toString();
  const hasAppliedEffect = status === 'applied' || originalStatus === 'applied';
  const clientId = adapter.request?.clientId ?? (await adapter.client.clientId());
  const result = {
    mutationId,
    status,
    checkpoint,
  };
  if (originalStatus !== undefined) result.originalStatus = originalStatus;
  if (hasAppliedEffect) result.revision = mutationId;

  return {
    protocolVersion: 1,
    clientId,
    lastMutationId: stateBigInt(state, 'response_watermark').toString(),
    checkpoint,
    results: [result],
  };
}

async function databaseSnapshot(client) {
  const [mutations, metadata] = await Promise.all([
    client.db.localMutations.orderBy('id').toArray(),
    client.db.meta.orderBy('key').toArray(),
  ]);
  return JSON.stringify({ mutations, metadata });
}

async function applyAction(adapter, state, tracePath) {
  if (state.action === 'init') {
    await destroyAdapter(adapter);
    return createAdapter(tracePath);
  }

  ensure(adapter !== null, `action \`${state.action}\` occurred before init`);

  switch (state.action) {
    case 'idle':
    case 'compact':
      break;

    case 'enqueue': {
      const expectedId = stateBigInt(state, 'next_id') - 1n;
      ensure(expectedId > 0n, 'enqueue produced an invalid next_id');
      const rowId = await adapter.client.queueMutation(
        'docs',
        `record-${expectedId}`,
        { id: `record-${expectedId}`, value: expectedId.toString() },
      );
      const row = await adapter.client.db.localMutations.get(rowId);
      ensure(row !== undefined, `queueMutation returned missing row ${rowId}`);
      ensure(
        row.mutationId === expectedId.toString(),
        `enqueue allocated ${row.mutationId ?? '<missing>'}, model allocated ${expectedId}`,
      );
      break;
    }

    case 'send': {
      ensure(adapter.request === null, 'request already in flight');
      ensure(adapter.response === null, 'response already present');
      const expectedId = pickedId(state);
      const request = await adapter.client.protocolPushRequest(1);
      ensure(
        requestMutationId(request) === expectedId,
        `sent mutation does not match model id ${expectedId}`,
      );
      adapter.request = request;
      break;
    }

    case 'apply_new':
    case 'reject_new': {
      ensure(adapter.response === null, 'response already present');
      ensure(
        requestMutationId(adapter.request) === pickedId(state),
        `${state.action} id does not match the in-flight request`,
      );
      adapter.response = await responseFromState(
        adapter,
        state,
        state.action === 'apply_new' ? 'applied' : 'rejected',
        undefined,
      );
      adapter.responseValid = true;
      break;
    }

    case 'reply_duplicate': {
      ensure(adapter.response === null, 'response already present');
      const id = pickedId(state);
      ensure(
        requestMutationId(adapter.request) === id,
        'duplicate reply id does not match the in-flight request',
      );
      const idString = id.toString();
      let originalStatus;
      if (stateSet(state, 'applied').has(idString)) originalStatus = 'applied';
      else if (stateSet(state, 'rejected').has(idString)) originalStatus = 'rejected';
      else throw invalid('duplicate reply has no durable original outcome');
      adapter.response = await responseFromState(
        adapter,
        state,
        'duplicate',
        originalStatus,
      );
      adapter.responseValid = true;
      break;
    }

    case 'inject_mismatched_response': {
      ensure(adapter.response === null, 'response already present');
      ensure(adapter.request !== null, 'mismatched response without an in-flight request');
      const response = await responseFromState(adapter, state, 'applied', undefined);
      const before = await databaseSnapshot(adapter.client);
      let rejected = false;
      try {
        await adapter.client.acknowledgePush(response, adapter.request);
      } catch (error) {
        rejected = true;
        ensure(
          error instanceof Error &&
            error.message === 'push acknowledgement does not match the sent batch',
          `malformed response returned unexpected error: ${error}`,
        );
      }
      ensure(rejected, 'the model-injected response must be rejected');
      ensure(
        (await databaseSnapshot(adapter.client)) === before,
        'rejecting a malformed response mutated IndexedDB state',
      );
      adapter.response = response;
      adapter.responseValid = false;
      break;
    }

    case 'lose_committed_response':
    case 'lose_uncommitted_request':
      adapter.request = null;
      adapter.response = null;
      adapter.responseValid = false;
      break;

    case 'discard_malformed_response':
      ensure(
        adapter.response !== null && !adapter.responseValid,
        'discard_malformed_response requires a rejected response',
      );
      adapter.request = null;
      adapter.response = null;
      adapter.responseValid = false;
      break;

    case 'acknowledge': {
      ensure(adapter.responseValid, 'cannot acknowledge an invalid response');
      ensure(adapter.response !== null, 'acknowledge without a response');
      ensure(adapter.request !== null, 'acknowledge without an in-flight request');
      ensure(
        requestMutationId(adapter.request) === pickedId(state),
        'acknowledgement id does not match the in-flight request',
      );
      const changed = await adapter.client.acknowledgePush(
        adapter.response,
        adapter.request,
      );
      ensure(changed === 1, `acknowledgement changed ${changed} rows`);
      adapter.request = null;
      adapter.response = null;
      adapter.responseValid = false;
      break;
    }

    case 'pull':
      await adapter.client.setPullCheckpoint(
        stateBigInt(state, 'local_checkpoint').toString(),
      );
      break;

    case 'begin_reset':
      ensure(!adapter.replacingSnapshot, 'snapshot replacement already active');
      adapter.replacingSnapshot = true;
      break;

    case 'crash_during_reset':
      ensure(
        adapter.replacingSnapshot,
        'crash_during_reset without an active replacement',
      );
      adapter.replacingSnapshot = false;
      break;

    case 'finish_reset': {
      ensure(adapter.replacingSnapshot, 'finish_reset without an active replacement');
      await adapter.client.installSnapshot(
        {
          protocolVersion: 1,
          checkpoint: stateBigInt(state, 'server_checkpoint').toString(),
          records: [],
        },
        async () => {},
      );
      adapter.replacingSnapshot = false;
      break;
    }

    default:
      throw invalid(`unsupported model action \`${state.action}\``);
  }

  return adapter;
}

async function assertProjection(adapter, state, context) {
  ensure(adapter !== null, `${context}: no adapter after action`);

  const sequence = (await adapter.client.db.meta.get(MUTATION_SEQUENCE_KEY))?.value ?? '0';
  ensure(/^(?:0|[1-9]\d*)$/.test(sequence), `${context}: invalid durable mutation sequence`);
  const actualNextId = BigInt(sequence) + 1n;
  const expectedNextId = stateBigInt(state, 'next_id');
  ensure(
    actualNextId === expectedNextId,
    `${context}: next id ${actualNextId} != model ${expectedNextId}`,
  );

  const expectedCheckpoint = stateBigInt(state, 'local_checkpoint').toString();
  const actualCheckpoint = await adapter.client.pullCheckpoint();
  ensure(
    actualCheckpoint === expectedCheckpoint,
    `${context}: checkpoint ${actualCheckpoint} != model ${expectedCheckpoint}`,
  );

  const [pendingRows, allRows] = await Promise.all([
    adapter.client.pendingMutations(),
    adapter.client.db.localMutations.orderBy('id').toArray(),
  ]);
  const actualPending = new Set(
    pendingRows.map((row) => {
      ensure(row.mutationId !== undefined, `${context}: pending row lacks mutation id`);
      return BigInt(row.mutationId).toString();
    }),
  );
  const actualConfirmed = new Set(
    allRows
      .filter((row) => row.syncStatus === SYNC_STATUS.SYNCED)
      .map((row) => {
        ensure(row.mutationId !== undefined, `${context}: confirmed row lacks mutation id`);
        return BigInt(row.mutationId).toString();
      }),
  );
  const actualAll = new Set(
    allRows.map((row) => {
      ensure(row.mutationId !== undefined, `${context}: allocated row lacks mutation id`);
      return BigInt(row.mutationId).toString();
    }),
  );

  ensureSetsEqual(actualPending, stateSet(state, 'pending'), context, 'pending');
  ensureSetsEqual(
    actualConfirmed,
    stateSet(state, 'acknowledged'),
    context,
    'confirmed',
  );
  ensureSetsEqual(actualAll, allocatedIds(expectedNextId), context, 'allocated ids');

  const inFlight = stateBigInt(state, 'in_flight');
  if (adapter.request === null) {
    ensure(inFlight === 0n, `${context}: model has in-flight id ${inFlight}, adapter has none`);
  } else {
    ensure(inFlight > 0n, `${context}: adapter has a request, model has no in-flight id`);
    ensure(
      requestMutationId(adapter.request) === inFlight,
      `${context}: in-flight request does not contain model id ${inFlight}`,
    );
  }

  const responsePresent = stateBool(state, 'response_present');
  ensure(
    (adapter.response !== null) === responsePresent,
    `${context}: response presence differs from model`,
  );
  if (adapter.response !== null) {
    const result = adapter.response.results[0];
    ensure(result !== undefined, `${context}: adapter response has no mutation result`);
    ensure(
      adapter.response.lastMutationId ===
        stateBigInt(state, 'response_watermark').toString(),
      `${context}: response watermark differs from model`,
    );
    ensure(
      adapter.response.checkpoint ===
        stateBigInt(state, 'response_checkpoint').toString(),
      `${context}: response checkpoint differs from model`,
    );
    ensure(
      result.mutationId === stateBigInt(state, 'response_mutation_id').toString(),
      `${context}: response mutation id differs from model`,
    );
    ensure(
      adapter.responseValid === stateBool(state, 'response_valid_for_in_flight'),
      `${context}: response validity differs from model`,
    );
  } else {
    ensure(!adapter.responseValid, `${context}: absent response cannot be marked valid`);
  }

  const replacing = stateTag(state, 'reset_phase') === 'Replacing';
  ensure(
    adapter.replacingSnapshot === replacing,
    `${context}: reset phase differs from model`,
  );
}

async function replay(tracePath) {
  const trace = JSON.parse(await readFile(tracePath, 'utf8'));
  ensure(Array.isArray(trace.states), `${tracePath}: ITF trace has no states array`);
  ensure(trace.states.length > 0, `${tracePath}: ITF trace has no states`);

  let adapter = null;
  try {
    for (let index = 0; index < trace.states.length; index += 1) {
      const state = trace.states[index];
      ensure(
        state !== null && typeof state === 'object' && typeof state.action === 'string',
        `${tracePath}: invalid state ${index}`,
      );
      const context = `${tracePath} state ${index} action ${state.action}`;
      try {
        adapter = await applyAction(adapter, state, tracePath);
        await assertProjection(adapter, state, context);
      } catch (error) {
        throw invalid(`${context}: ${error instanceof Error ? error.message : String(error)}`);
      }
    }
    return trace.states.length;
  } finally {
    await destroyAdapter(adapter);
  }
}

async function main() {
  const paths = process.argv.slice(2).sort();
  ensure(
    paths.length > 0,
    'usage: node formal/quint-itf-replay.mjs <trace.itf.json>...',
  );

  let states = 0;
  for (const tracePath of paths) {
    const count = await replay(tracePath);
    states += count;
    process.stdout.write(`replayed ${count} model states from ${tracePath}\n`);
  }
  process.stdout.write(
    `TypeScript OptoSyncClient conformed to ${states} states across ${paths.length} Quint ITF traces\n`,
  );
}

main().catch((error) => {
  process.stderr.write(`${error instanceof Error ? error.stack ?? error.message : String(error)}\n`);
  process.exitCode = 1;
});

File: clients/rust/examples/quint_itf_replay.rs
//! Replay Quint's Informal Trace Format (ITF) model-based traces against the
//! production Rust [`ProtocolQueue`] state machine.
//!
//! This adapter deliberately compares only observable client state. Server-only
//! model fields (ledger, effect count, retention floor) are used to synthesize
//! protocol responses, while queue allocation, request construction,
//! acknowledgement rejection/acceptance, checkpoint installation, and reset
//! durability execute through the real library API.

use opto_sync_client::protocol::{
    LocalProtocolStatus, MutationResult, ProtocolError, ProtocolQueue, PushRequest, PushResponse,
    ResultStatus, SnapshotResponse,
};
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::BTreeSet;
use std::error::Error;
use std::fmt::Display;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

const CLIENT_ID: &str = "formal-client";

type AnyResult<T> = Result<T, Box<dyn Error>>;

#[derive(Debug, Deserialize)]
struct ItfTrace {
    states: Vec<ItfState>,
}

#[derive(Debug, Deserialize)]
struct ItfState {
    #[serde(rename = "mbt::actionTaken")]
    action: String,
    #[serde(rename = "mbt::nondetPicks", default)]
    nondet_picks: Value,
    s: Value,
}

struct Adapter {
    queue: ProtocolQueue,
    request: Option<PushRequest>,
    response: Option<PushResponse>,
    response_valid: bool,
    replacing_snapshot: bool,
}

impl Adapter {
    fn new() -> AnyResult<Self> {
        Ok(Self {
            queue: ProtocolQueue::new(CLIENT_ID)?,
            request: None,
            response: None,
            response_valid: false,
            replacing_snapshot: false,
        })
    }
}

fn invalid(message: impl Into<String>) -> Box<dyn Error> {
    Box::new(io::Error::new(io::ErrorKind::InvalidData, message.into()))
}

fn ensure(condition: bool, message: impl Into<String>) -> AnyResult<()> {
    if condition {
        Ok(())
    } else {
        Err(invalid(message))
    }
}

fn field<'a>(value: &'a Value, name: &str) -> AnyResult<&'a Value> {
    value
        .get(name)
        .ok_or_else(|| invalid(format!("missing ITF field `{name}`")))
}

fn tagged_bigint(value: &Value) -> AnyResult<u64> {
    field(value, "#bigint")?
        .as_str()
        .ok_or_else(|| invalid("ITF #bigint must contain a decimal string"))?
        .parse::<u64>()
        .map_err(|error| invalid(format!("invalid ITF #bigint: {error}")))
}

fn state_u64(state: &ItfState, name: &str) -> AnyResult<u64> {
    tagged_bigint(field(&state.s, name)?)
}

fn state_bool(state: &ItfState, name: &str) -> AnyResult<bool> {
    field(&state.s, name)?
        .as_bool()
        .ok_or_else(|| invalid(format!("ITF state field `{name}` must be boolean")))
}

fn state_set(state: &ItfState, name: &str) -> AnyResult<BTreeSet<u64>> {
    let entries = field(field(&state.s, name)?, "#set")?
        .as_array()
        .ok_or_else(|| invalid(format!("ITF state field `{name}` must be a set")))?;
    entries.iter().map(tagged_bigint).collect()
}

fn state_tag<'a>(state: &'a ItfState, name: &str) -> AnyResult<&'a str> {
    field(field(&state.s, name)?, "tag")?
        .as_str()
        .ok_or_else(|| invalid(format!("ITF state field `{name}` must be tagged")))
}

fn picked_id(state: &ItfState) -> AnyResult<u64> {
    let pick = field(&state.nondet_picks, "id")?;
    ensure(
        field(pick, "tag")?.as_str() == Some("Some"),
        format!("action `{}` requires a nondeterministic id", state.action),
    )?;
    tagged_bigint(field(pick, "value")?)
}

fn response_from_state(
    state: &ItfState,
    status: ResultStatus,
    original_status: Option<ResultStatus>,
) -> AnyResult<PushResponse> {
    let mutation_id = state_u64(state, "response_mutation_id")?.to_string();
    let checkpoint = state_u64(state, "response_checkpoint")?.to_string();
    let has_applied_effect =
        status == ResultStatus::Applied || original_status == Some(ResultStatus::Applied);

    Ok(PushResponse {
        protocol_version: 1,
        client_id: CLIENT_ID.to_string(),
        last_mutation_id: state_u64(state, "response_watermark")?.to_string(),
        checkpoint: checkpoint.clone(),
        results: vec![MutationResult {
            mutation_id: mutation_id.clone(),
            status,
            original_status,
            checkpoint: Some(checkpoint),
            revision: has_applied_effect.then_some(mutation_id),
            code: None,
            message: None,
        }],
    })
}

fn request_mutation_id(request: &PushRequest) -> AnyResult<u64> {
    ensure(
        request.mutations.len() == 1,
        "formal adapter sends exactly one mutation per request",
    )?;
    request.mutations[0]
        .mutation_id
        .parse::<u64>()
        .map_err(|error| invalid(format!("invalid request mutation id: {error}")))
}

fn apply_action(adapter: &mut Adapter, state: &ItfState) -> AnyResult<()> {
    match state.action.as_str() {
        "init" => {
            *adapter = Adapter::new()?;
        }
        "idle" | "compact" => {}
        "enqueue" => {
            let expected_id = state_u64(state, "next_id")?
                .checked_sub(1)
                .ok_or_else(|| invalid("enqueue produced an invalid next_id"))?;
            let actual_id = adapter.queue.queue_upsert(
                "docs",
                format!("record-{expected_id}"),
                json!({"id": format!("record-{expected_id}"), "value": expected_id}),
                None,
                false,
            )?;
            ensure(
                actual_id == expected_id.to_string(),
                format!("enqueue allocated {actual_id}, model allocated {expected_id}"),
            )?;
        }
        "send" => {
            ensure(adapter.request.is_none(), "request already in flight")?;
            ensure(adapter.response.is_none(), "response already present")?;
            let expected_id = picked_id(state)?;
            let request = adapter.queue.push_request(1)?;
            ensure(
                request_mutation_id(&request)? == expected_id,
                format!("sent mutation does not match model id {expected_id}"),
            )?;
            adapter.request = Some(request);
        }
        "apply_new" => {
            ensure(adapter.response.is_none(), "response already present")?;
            ensure(
                request_mutation_id(
                    adapter
                        .request
                        .as_ref()
                        .ok_or_else(|| invalid("apply_new without an in-flight request"))?,
                )? == picked_id(state)?,
                "apply_new id does not match the in-flight request",
            )?;
            adapter.response = Some(response_from_state(state, ResultStatus::Applied, None)?);
            adapter.response_valid = true;
        }
        "reject_new" => {
            ensure(adapter.response.is_none(), "response already present")?;
            ensure(
                request_mutation_id(
                    adapter
                        .request
                        .as_ref()
                        .ok_or_else(|| invalid("reject_new without an in-flight request"))?,
                )? == picked_id(state)?,
                "reject_new id does not match the in-flight request",
            )?;
            adapter.response = Some(response_from_state(state, ResultStatus::Rejected, None)?);
            adapter.response_valid = true;
        }
        "reply_duplicate" => {
            ensure(adapter.response.is_none(), "response already present")?;
            let id = picked_id(state)?;
            ensure(
                request_mutation_id(
                    adapter
                        .request
                        .as_ref()
                        .ok_or_else(|| invalid("duplicate reply without an in-flight request"))?,
                )? == id,
                "duplicate reply id does not match the in-flight request",
            )?;
            let original_status = if state_set(state, "applied")?.contains(&id) {
                ResultStatus::Applied
            } else if state_set(state, "rejected")?.contains(&id) {
                ResultStatus::Rejected
            } else {
                return Err(invalid("duplicate reply has no durable original outcome"));
            };
            adapter.response = Some(response_from_state(
                state,
                ResultStatus::Duplicate,
                Some(original_status),
            )?);
            adapter.response_valid = true;
        }
        "inject_mismatched_response" => {
            ensure(adapter.response.is_none(), "response already present")?;
            let response = response_from_state(state, ResultStatus::Applied, None)?;
            let request = adapter
                .request
                .as_ref()
                .ok_or_else(|| invalid("mismatched response without an in-flight request"))?;
            let before = serde_json::to_value(&adapter.queue)?;
            let error = adapter
                .queue
                .acknowledge(&response, request)
                .expect_err("the model-injected response must be rejected");
            ensure(
                error == ProtocolError::InvalidAcknowledgement,
                format!("malformed response returned unexpected error: {error}"),
            )?;
            ensure(
                serde_json::to_value(&adapter.queue)? == before,
                "rejecting a malformed response mutated the queue",
            )?;
            adapter.response = Some(response);
            adapter.response_valid = false;
        }
        "lose_committed_response" | "lose_uncommitted_request" => {
            adapter.request = None;
            adapter.response = None;
            adapter.response_valid = false;
        }
        "discard_malformed_response" => {
            ensure(
                adapter.response.is_some() && !adapter.response_valid,
                "discard_malformed_response requires a rejected response",
            )?;
            adapter.request = None;
            adapter.response = None;
            adapter.response_valid = false;
        }
        "acknowledge" => {
            ensure(
                adapter.response_valid,
                "cannot acknowledge an invalid response",
            )?;
            let response = adapter
                .response
                .as_ref()
                .ok_or_else(|| invalid("acknowledge without a response"))?;
            let request = adapter
                .request
                .as_ref()
                .ok_or_else(|| invalid("acknowledge without an in-flight request"))?;
            ensure(
                request_mutation_id(request)? == picked_id(state)?,
                "acknowledgement id does not match the in-flight request",
            )?;
            let changed = adapter.queue.acknowledge(response, request)?;
            ensure(
                changed == 1,
                format!("acknowledgement changed {changed} rows"),
            )?;
            adapter.request = None;
            adapter.response = None;
            adapter.response_valid = false;
        }
        "pull" => {
            adapter
                .queue
                .set_checkpoint(state_u64(state, "local_checkpoint")?.to_string())?;
        }
        "begin_reset" => {
            ensure(
                !adapter.replacing_snapshot,
                "snapshot replacement already active",
            )?;
            adapter.replacing_snapshot = true;
        }
        "crash_during_reset" => {
            ensure(
                adapter.replacing_snapshot,
                "crash_during_reset without an active replacement",
            )?;
            adapter.replacing_snapshot = false;
        }
        "finish_reset" => {
            ensure(
                adapter.replacing_snapshot,
                "finish_reset without an active replacement",
            )?;
            let snapshot = SnapshotResponse {
                protocol_version: 1,
                checkpoint: state_u64(state, "server_checkpoint")?.to_string(),
                records: Vec::new(),
            };
            adapter
                .queue
                .install_snapshot(&snapshot, |_| Ok::<(), &'static str>(()))?;
            adapter.replacing_snapshot = false;
        }
        action => return Err(invalid(format!("unsupported model action `{action}`"))),
    }
    Ok(())
}

fn parse_mutation_id(value: impl Display) -> AnyResult<u64> {
    value
        .to_string()
        .parse::<u64>()
        .map_err(|error| invalid(format!("invalid queue mutation id: {error}")))
}

fn assert_projection(adapter: &Adapter, state: &ItfState, context: &str) -> AnyResult<()> {
    adapter.queue.validate()?;

    let queue_json = serde_json::to_value(&adapter.queue)?;
    let actual_next_id = field(&queue_json, "nextMutationId")?
        .as_u64()
        .ok_or_else(|| invalid("serialized nextMutationId must be an unsigned integer"))?;
    let expected_next_id = state_u64(state, "next_id")?;
    ensure(
        actual_next_id == expected_next_id,
        format!("{context}: next id {actual_next_id} != model {expected_next_id}"),
    )?;

    let expected_checkpoint = state_u64(state, "local_checkpoint")?.to_string();
    ensure(
        adapter.queue.checkpoint() == expected_checkpoint,
        format!(
            "{context}: checkpoint {} != model {expected_checkpoint}",
            adapter.queue.checkpoint()
        ),
    )?;

    let actual_pending = adapter
        .queue
        .pending()
        .map(|mutation| parse_mutation_id(&mutation.mutation_id))
        .collect::<AnyResult<BTreeSet<_>>>()?;
    let actual_confirmed = adapter
        .queue
        .all()
        .iter()
        .filter(|mutation| mutation.status == LocalProtocolStatus::Confirmed)
        .map(|mutation| parse_mutation_id(&mutation.mutation_id))
        .collect::<AnyResult<BTreeSet<_>>>()?;
    let expected_pending = state_set(state, "pending")?;
    let expected_confirmed = state_set(state, "acknowledged")?;
    ensure(
        actual_pending == expected_pending,
        format!("{context}: pending {actual_pending:?} != model {expected_pending:?}"),
    )?;
    ensure(
        actual_confirmed == expected_confirmed,
        format!("{context}: confirmed {actual_confirmed:?} != model {expected_confirmed:?}"),
    )?;

    let actual_all = adapter
        .queue
        .all()
        .iter()
        .map(|mutation| parse_mutation_id(&mutation.mutation_id))
        .collect::<AnyResult<BTreeSet<_>>>()?;
    let expected_all = (1..expected_next_id).collect::<BTreeSet<_>>();
    ensure(
        actual_all == expected_all,
        format!("{context}: allocated ids {actual_all:?} != model {expected_all:?}"),
    )?;

    let in_flight = state_u64(state, "in_flight")?;
    match (&adapter.request, in_flight) {
        (None, 0) => {}
        (Some(request), id) if id > 0 => ensure(
            request_mutation_id(request)? == id,
            format!("{context}: in-flight request does not contain model id {id}"),
        )?,
        (None, id) => {
            return Err(invalid(format!(
                "{context}: model has in-flight id {id}, adapter has no request"
            )))
        }
        (Some(_), 0) => {
            return Err(invalid(format!(
                "{context}: adapter has a request, model has no in-flight id"
            )))
        }
        (Some(_), _) => unreachable!("covered positive in-flight id"),
    }

    let response_present = state_bool(state, "response_present")?;
    ensure(
        adapter.response.is_some() == response_present,
        format!("{context}: response presence differs from model"),
    )?;
    if let Some(response) = &adapter.response {
        let result = response
            .results
            .first()
            .ok_or_else(|| invalid("adapter response has no mutation result"))?;
        ensure(
            response.last_mutation_id == state_u64(state, "response_watermark")?.to_string(),
            format!("{context}: response watermark differs from model"),
        )?;
        ensure(
            response.checkpoint == state_u64(state, "response_checkpoint")?.to_string(),
            format!("{context}: response checkpoint differs from model"),
        )?;
        ensure(
            result.mutation_id == state_u64(state, "response_mutation_id")?.to_string(),
            format!("{context}: response mutation id differs from model"),
        )?;
        ensure(
            adapter.response_valid == state_bool(state, "response_valid_for_in_flight")?,
            format!("{context}: response validity differs from model"),
        )?;
    } else {
        ensure(
            !adapter.response_valid,
            format!("{context}: absent response cannot be marked valid"),
        )?;
    }

    let replacing = state_tag(state, "reset_phase")? == "Replacing";
    ensure(
        adapter.replacing_snapshot == replacing,
        format!("{context}: reset phase differs from model"),
    )?;

    Ok(())
}

fn replay(path: &Path) -> AnyResult<usize> {
    let bytes = fs::read(path)?;
    let trace: ItfTrace = serde_json::from_slice(&bytes)?;
    ensure(!trace.states.is_empty(), "ITF trace has no states")?;

    let mut adapter = Adapter::new()?;
    for (index, state) in trace.states.iter().enumerate() {
        let context = format!("{} state {index} action {}", path.display(), state.action);
        apply_action(&mut adapter, state)
            .map_err(|error| invalid(format!("{context}: {error}")))?;
        assert_projection(&adapter, state, &context)?;
    }
    Ok(trace.states.len())
}

fn main() -> AnyResult<()> {
    let mut paths = std::env::args_os()
        .skip(1)
        .map(PathBuf::from)
        .collect::<Vec<_>>();
    ensure(
        !paths.is_empty(),
        "usage: cargo run --example quint_itf_replay -- <trace.itf.json>...",
    )?;
    paths.sort();

    let mut states = 0usize;
    for path in &paths {
        let count = replay(path)?;
        states += count;
        println!("replayed {count} model states from {}", path.display());
    }
    println!(
        "Rust ProtocolQueue conformed to {states} states across {} Quint ITF traces",
        paths.len()
    );
    Ok(())
}
